docs/design/chart_format.md

```diff
@@ -0,0 +1,664 @@
+
+```
+//helm:generate TOOL [ARGS]\n
+#helm:generate TOOL [ARGS]\n
+/*helm:generate TOOL [ARGS]*/
+```
+
+`TOOL` corresponds to a registered tool name (see [this proposal](https://github.com/kubernetes/kubernetes/pull/18215#issuecomment-168885511)). `[ARGS]` are zero or more space-separated strings that will be passed unmodified to the tool. The first two formats require a terminating line feed. The third format requires a terminating `*/` sequence.
+
+
+**Example:**
+
+```
+#helm:generate dm_template
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+
+In the example above, the `#helm:generate dm_template` indicates that the remainder of the file contents should be processed using whatever handler is registered for handling `dm_template`.
+
+
+Generator tools may take arguments. No format is specified for arguments. For example, one might use UNIX-like flags.
+
+
+```
+#helm:generate dm_template --ignore-timezone -s
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+Any arguments are passed into the generator unaltered. Generators may ignore arguments that they do not support or understand.
+
+
+A generator tool _must_ not overwrite any files (including itself) present in the chart’s `templates/` directory. It must not alter the `README.md`, the `LICENSE`, the `Chart.yaml`, or the contents of the `hooks/` directory. These restrictions maintain the reproducibility of the build over subsequent generator runs.
+
+
+One common form of generator tool is a template renderer. Such a tool takes a template (and possibly other supporting files) as input, processes the template, and outputs one or more generated files or data streams. The location of output is not specified by this proposal (though it is subject to the caveats above).
+
+
+An implementation of this specification may support any number of template engines. However, two are explicitly defined: Python/Jinja templates (formerly DM Templates), and simple Go templates (formerly Helm Templates).
+
+#### Python/Jinja Templates (aka DM Templates)
+Python/Jinja templates, [defined here](https://github.com/kubernetes/deployment-manager/blob/master/docs/design/design.md), have been supported since DM’s inception. The only amendment this proposal makes to the existing format is that template and Python files must have `helm:generate` headers to be processed. (Schema files do not need generate headers because no operations are run directly on these files.)
+
+
+The reserved generator name for such templates is `dm_template`.
```

Comment:
We don't need to reserve a generator name if we follow the proposal referenced above. dm will be registered by default with a well known name, but users _could_ choose to register it under a different one. helm doctor could check that there's a registered generator for the name used in a given chart.